### PR TITLE
Add async loading to reduce package size

### LINKED ISSUE #28: Async loading of dash-vtk for multi-page apps
Since the bundle for dash-vtk is quite large (1MB), it might be beneficial for multi-page apps (such as dash-docs) if dash-vtk is loaded asynchronously, such that people that do not visit the page(s) featuring dash-vtk don't have to load the bundle.

I think this PR shows how to add async components: https://github.com/plotly/dash-core-components/pull/739

^^ @chriddyp correct me if it's not the correct PR

### PR DESCRIPTION
**Pull request by xhlulu**
closes #28 

More information in PR https://github.com/plotly/dash-vtk/pull/40

Screenshot of network when `dash_vtk` is imported but not used:
image

Pasting some explanation of the `asyncComponentBuilder` function that I shared with @Marc-Andre-Rivet 

> ```js
> const AsyncComponentBuilder = name => async () => {
>     const AsyncReactVTK = import(/*webpackChunkName: "ReactVTK" */ './ReactVTK');
>     const ReactVTK = await AsyncReactVTK;
>     // console.log("ReactVTK", ReactVTK);
>     // window.ReactVTK = ReactVTK;
>     const reactComponent = ReactVTK.default[name];
>     // We need to trick React.lazy into thinking we are giving
>     // the output of an import("my-module.js") Promise.
>     const fakeModule = {default: reactComponent};
>     return fakeModule;
> }
> ```
> Let's look at individual lines:
> 
> ```
> const AsyncReactVTK = import(/*webpackChunkName: "ReactVTK" */ './ReactVTK');
> ```
> above i'm using the asynchronously importing the "react-vtk-js" library (you can ignore ./ReactVTK, it just imports * then exports the react-vtk-js library).
> ```
>  const ReactVTK = await AsyncReactVTK;
> ```
> above, Just resolving the promise here, so ReactVTK is an actual module instead of a promise.
> ```
> const reactComponent = ReactVTK.default[name];
> ```
> above i'm accessing the react component by the name, so if name="Algorithm" then it'd be equivalent to `ReactVTK.default.Algorithm`. Usually it would simply be `ReactVTK.Algorithm` but we need the default due to how i export `react-vtk-js` in `./ReactVTK.js` .
> ```
> const fakeModule = {default: reactComponent};
> ```
> Just as the comment says: `React.lazy` expects a module, we give it an object that behaves like a module.
> ```
> const AsyncComponentBuilder = name => async () => {...}
> ```
> above, `AsyncComponentBuilder` is just a HOF that given a name (e.g. "View" or "Algorithm") will return an async function that we can give to `React.lazy`  and use inside `React.Suspense`.

### CODE REVIEW

File: src/lib/components/View.react.js [hunk 1]
```diff
@@ -2,7 +2,7 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 
-import { View as VtkItem } from 'react-vtk-js';
+import VtkItem from 'react-vtk-js/dist/esm/core/View';
```
Comment (jourdain) [inline src/lib/components/View.react.js:5]: A better import should be `import VtkItem from 'react-vtk-js/core/View.js';`

Comment (xhlulu) [reply src/lib/components/View.react.js:5]: I don't think that works

Comment (xhlulu) [reply src/lib/components/View.react.js:5]:
```
Critical dependency: the request of a dependency is an expression
 @ ./node_modules/react-vtk-js/dist/esm/node_modules/vtk.js/dist/esm/vendor/readable-stream/lib/_stream_duplex.js
 @ ./node_modules/react-vtk-js/dist/esm/node_modules/vtk.js/dist/esm/Filters/Sources/ConcentricCylinderSource.js
 @ ./node_modules/react-vtk-js/dist/esm/index.js
 @ ./src/lib/components/Algorithm.react.js
 @ ./src/lib/index.js

ERROR in ./src/lib/components/View.react.js
Module not found: Error: Can't resolve 'react-vtk-js/core/View' in '/home/xhlu/dev/dash-vtk/src/lib/components'
 @ ./src/lib/components/View.react.js 4:0-45 15:42-49
 @ ./src/lib/index.js
npm ERR! code ELIFECYCLE
npm ERR! errno 2
```

**Comment by xhlulu** [comment]
@alexcjohnson @Marc-Andre-Rivet I pushed some commits here using Suspense and lazy. Let me know if that's what I'm supposed to do; if there's no problem I'll do the same thing for the other components.

File: dash_vtk/package-info.json [hunk 2]
```diff
@@ -30,6 +30,7 @@
   "devDependencies": {
     "@babel/core": "^7.5.4",
     "@babel/plugin-proposal-object-rest-spread": "^7.5.4",
+    "@babel/plugin-syntax-dynamic-import": "^7.8.3",
```
Comment (Marc-Andre-Rivet) [inline dash_vtk/package-info.json:33]: While necessary, using the babel dynamic import plugin is not enough. You also need to use https://www.npmjs.com/package/@plotly/webpack-dash-dynamic-import and to add that plugin in your webpack config file like so https://github.com/plotly/dash-core-components/blob/dev/webpack.config.js#L134

File: package.json [hunk 3]
```diff
@@ -30,6 +30,7 @@
   "devDependencies": {
     "@babel/core": "^7.5.4",
     "@babel/plugin-proposal-object-rest-spread": "^7.5.4",
+    "@babel/plugin-syntax-dynamic-import": "^7.8.3",
```
Comment (Marc-Andre-Rivet) [inline package.json:33]: While necessary, using the babel dynamic import plugin is not enough. You also need to use https://www.npmjs.com/package/@plotly/webpack-dash-dynamic-import and to add that plugin in your webpack config file like so https://github.com/plotly/dash-core-components/blob/dev/webpack.config.js#L134

File: src/lib/components/View.react.js (hunk 1 above)
Comment (Marc-Andre-Rivet) [reply src/lib/components/View.react.js]: Import the usual `react-vtk-js` instead. The way these esm modules are published isn't compatible with our usage. You can see above that the code is looking for `./node_modules/react-vtk-js/dist/esm/node_modules/vtk.js` but that folder does not exist in the released package, see: https://unpkg.com/browse/react-vtk-js@1.2.1/dist/esm/node_modules/

Comment (xhlulu) [reply src/lib/components/View.react.js]: @jourdain ^^

Comment (jourdain) [reply src/lib/components/View.react.js]: I'll see what I can do to move the required devDep of vtk.js into dep.

Comment (jourdain) [reply src/lib/components/View.react.js]: Just pushed `react-vtk-js@1.2.2` for you to try out. Normally the esm bundle should have all its dependency.

**Comment by xhlulu** [comment]
@Marc-Andre-Rivet Thanks for the advice. I had a pair-coding session with Joe and I was succefully in wrapping one of the components (`dash_vtk.View`) using the chunked async approach; unfortunately I started running into problems when I try to add async to the other 19 components; more about that in the slack channel `#dash-vtk`.

**Comment by xhlulu** [comment]
I'm rerolling back to before adding async so I can explore the single-chunk async apporach (load all of vtk or none).

I'm saving the state of this branch here: https://github.com/plotly/dash-vtk/tree/lazy-loading-archive-1

**Comment by xhlulu** [comment]
@alexcjohnson @Marc-Andre-Rivet @jdamiba This PR is ready to be reviewed! I think the async should work now, and all the problem on the vtk side have been resolved by @jourdain . Feel free to take a look at the percy snapshot to validate the result.

**Review by jdamiba — APPROVED** [review]
This looks good to me as well. 💃🏿 

I tested this locally by running the `usage.py` example both with and without `dash-vtk` components in the app layout, and can confirm that the async `ReactVTK.js` bundle is only served when `dash-vtk` components are in the layout. When the library is simply imported into an app, only the thin shell is loaded. 

The same holds true at https://dash-docs-pr-1091.herokuapp.com/ and https://dash-docs-pr-1091.herokuapp.com/vtk, which are a version of the `dash-docs` which point to this branch of the `dash-vtk` library.

File: dash_vtk/package-info.json [hunk 4]
```diff
@@ -24,14 +24,17 @@
   "author": "Kitware Inc <sebastien.jourdain@kitware.com> and Plotly Technologies <xinghan@plot.ly>",
   "license": "MIT",
   "dependencies": {
+    "@plotly/dash-component-plugins": "^1.2.0",
     "ramda": "^0.26.1",
-    "react-vtk-js": "1.2.1"
+    "react-vtk-js": "^1.4.2"
```
Comment (Marc-Andre-Rivet) [inline dash_vtk/package-info.json:29]: Latest now at 1.5.0?

Comment (jourdain) [reply dash_vtk/package-info.json:29]: yes

File: webpack.config.js [hunk 5]
```diff
@@ -46,21 +50,38 @@ module.exports = (env, argv) => {
...
             path: path.resolve(__dirname, dashLibraryName),
+            chunkFilename: '[name].js',
             filename,
             library: dashLibraryName,
             libraryTarget: 'window',
         },
         devtool,
         externals,
         module: {
+            // noParse: /node_modules[\\\/]plotly.js/,
```
Comment (Marc-Andre-Rivet) [inline webpack.config.js:61]: 🔪

File: webpack.config.js [hunk 6]
```diff
@@ -46,21 +50,38 @@ module.exports = (env, argv) => {
...
                 {
                     test: /\.jsx?$/,
                     exclude: /node_modules/,
+                    use: {
+                        loader: 'babel-loader'
+                    }
+                },
+                {
+                    test: /\.jsx?$/,
+                    include: /node_modules[\\\/](react-jsx-parser|highlight[.]js)[\\\/]/,
```
Comment (Marc-Andre-Rivet) [inline webpack.config.js:72]: 🔪 - don't keep the 🍝 that's not needed :)

File: webpack.config.js [hunk 7]
```diff
@@ -77,5 +98,43 @@ module.exports = (env, argv) => {
...
+                            return `${cacheGroupKey}-${chunks[0].name}`;
+                        }
+                    },
+                    shared: {
+                        chunks: 'all',
+                        minSize: 0,
+                        minChunks: 2,
+                        name: 'dash_vtk-shared'
+                    }
+                }
```
Comment (Marc-Andre-Rivet) [inline webpack.config.js:129]: 🔪 shared - not being used as currently setup

File: src/lib/AsyncReactVTK.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ */
+const asyncComponentBuilder = name => async () => {
+    // eslint-disable-next-line
+    const AsyncReactVTK = import(/* webpackChunkName: "ReactVTK" */ './ReactVTK');
+    const ReactVTK = await AsyncReactVTK;
+
+    const reactComponent = ReactVTK.default[name];
+    const fakeModule = { default: reactComponent };
+    return fakeModule;
+}
```
Comment (Marc-Andre-Rivet) [inline src/lib/AsyncReactVTK.js:15]: Doesn't need to be changed but could be simplified to something like:

```
const asncComponentBuilder = name => import(/* webpackChunkName: "ReactVTK" */ './ReactVTK').then(ReactVTK => ({ default: ReactVTK.default[name] }))
```

Comment (xhlulu) [reply src/lib/AsyncReactVTK.js:15]: The code snippet doesn't seem to work:

image

Comment (xhlulu) [reply src/lib/AsyncReactVTK.js:15]: Ok so this should work:

```js
const asyncComponentBuilder = name => () => import(/* webpackChunkName: "ReactVTK" */ './ReactVTK').then(ReactVTK => ({ default: ReactVTK.default[name] }))
```

I believe it achieves the same thing in a different way, but it comes down to whether to use `Promise.prototype.then()` or `async`/`await`.

Comment (jdamiba) [reply src/lib/AsyncReactVTK.js:15]: I would prefer to keep this code as is. Thinking about looking at this code in 6 months after forgetting the details of the implementation, I think I'm more likely to be able to read through the current implementation line by line and follow the train of transformations than I am to be able to parse the one-liner.

Comment (Marc-Andre-Rivet) [reply src/lib/AsyncReactVTK.js:15]: To be honest I disagree as to what is most readable here but as previously stated I'm fine with the current implementation. Functionally both are equivalent.

**Review by Marc-Andre-Rivet — APPROVED** [review]
💃 - @xhlulu, @jdamiba good job digging through all the details to make this (yet to be documented feature) work for this package 👍

**Comment by xhlulu** [comment]
Thank you @Marc-Andre-Rivet @jdamiba @jourdain! Merging this now.

**Pull request merged**